src/libData/BlockChainData/BlockChain.h

```diff
@@ -0,0 +1,176 @@
+
+/// Transient storage for DS/Tx/VC blocks.
+template<class T> class BlockChain
+{
+    std::mutex m_mutexBlocks;
+    CircularArray<T> m_blocks;
+
+public:
+    /// Constructor.
+    BlockChain() { Reset(); }
+
+    /// Destructor.
+    ~BlockChain() {}
+
+    /// Reset
+    void Reset() { m_blocks.resize(BLOCKCHAIN_SIZE); }
+
+    /// Returns the number of blocks.
+    boost::multiprecision::uint256_t GetBlockCount()
+    {
+        lock_guard<mutex> g(m_mutexBlocks);
+        return m_blocks.size();
+    }
+
+    /// Returns the last stored block.
+    T GetLastBlock()
+    {
+        lock_guard<mutex> g(m_mutexBlocks);
+        return m_blocks.back();
+    }
+
+    /// Returns the block at the specified block number.
+    TxBlock GetTxBlock(const boost::multiprecision::uint256_t& blockNum)
+    {
+        lock_guard<mutex> g(m_mutexBlocks);
+
+        if (blockNum >= m_blocks.size())
+        {
+            throw "Blocknumber Absent";
+        }
+        else if (blockNum + m_blocks.capacity() < m_blocks.size())
+        {
+            TxBlockSharedPtr block;
+            BlockStorage::GetBlockStorage().GetTxBlock(blockNum, block);
+            return *block;
+        }
+
+        // To-do: We cannot even index into a vector using uint256_t
+        // uint256_t might just be too big to begin with
+        // Consider switching to uint64_t
+        // For now we directly cast to uint64_t
+
+        if (m_blocks[blockNum].GetHeader().GetBlockNum() != blockNum)
+        {
+            LOG_GENERAL(FATAL,
+                        "assertion failed (" << __FILE__ << ":" << __LINE__
+                                             << ": " << __FUNCTION__ << ")");
+        }
+
+        return m_blocks[blockNum];
+    }
+
+    DSBlock GetDSBlock(const boost::multiprecision::uint256_t& blockNum)
+    {
+        lock_guard<mutex> g(m_mutexBlocks);
+
+        if (blockNum >= m_blocks.size())
+        {
+            throw "Blocknumber Absent";
+        }
+        else if (blockNum + m_blocks.capacity() < m_blocks.size())
+        {
+            DSBlockSharedPtr block;
+            BlockStorage::GetBlockStorage().GetDSBlock(blockNum, block);
+            return *block;
+        }
+
+        // To-do: We cannot even index into a vector using uint256_t
+        // uint256_t might just be too big to begin with
+        // Consider switching to uint64_t
+        // For now we directly cast to uint64_t
+
+        if (m_blocks[blockNum].GetHeader().GetBlockNum() != blockNum)
+        {
+            LOG_GENERAL(FATAL,
+                        "assertion failed (" << __FILE__ << ":" << __LINE__
+                                             << ": " << __FUNCTION__ << ")");
+        }
+
+        return m_blocks[blockNum];
+    }
+
+    VCBlock GetVCBlock(const boost::multiprecision::uint256_t& blockNum)
+    {
+        lock_guard<mutex> g(m_mutexBlocks);
+
+        if (blockNum >= m_blocks.size())
+        {
+            throw "Blocknumber Absent";
```

Comment:
I think throwing a string is not an ideal way to handle this. Maybe we can throw nullptr or log warning or log fatal in this case